Add this variable for use by automation and script templates

r-t-s opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a "this" variable to be used in templates for automation and scripts.
In automation "this" contains the states of the automation entity when rendering automation variables and trigger variables.
In script "this" contains the states of the script entity when rendering the script variables.

Note: This replaces #52556

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: [18505](https://github.com/home-assistant/home-assistant.io/pull/18505)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing 

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

r-t-s committed:
Add this variable for use by automation and script templates

yuvalabou commented:
What is the use case? Automations and scripts have binary states which are on or off, and if 'on' they will work and off otherwise.. and if they won't work we can't use the 'this' state?

r-t-s commented:
The use case is not really for the state but for the additional information that states contain (i.e. automation enitity_id and friendly_name)

Use Cases:

1. A blueprint instantiated automation want's to issue a notification of which "automation" is active.  At present this is not possible.  Now you can use  "{{ this.attributes.friendly_name}}"
2. In blueprints instantiated automations, allocate resources that are uniquely identified and used by each instance of the automation. 

Currently I have a blueprint that needs a unique  input_text and timer for each instance of the blueprint. I have to manually configure these and bind them via inputs in the blueprint.

I have developed an integration that allows these to be dynamically configured at startup time of the automation instance. But each automation instance needs a uniquely identified timer and input_text.  I can use the entity_id of the automation to dynamically generate the entity_id of the dynamically configured input_text and timer. Now when I want to instantiate the automation, I no longer need to manually configure the required resources first. 

Automations and Scripts are very similar, "this" was added to scripts to maintain parity.

emontnemery commented on homeassistant/components/automation/__init__.py:
```diff
@@ -460,7 +460,11 @@ async def async_trigger(self, run_variables, context=None, skip_condition=False)
                 try:
-                    variables = self._variables.async_render(self.hass, run_variables)
+                    states = self.hass.states.get(self.entity_id)
+                    this = states.as_dict() if states else {"entity_id": self.entity_id}
```
Can the state really be `None` when the automation runs? Did you manage to provoke that somehow?

emontnemery commented on homeassistant/components/automation/__init__.py:
```diff
@@ -585,8 +589,12 @@ def log_cb(level, msg, **kwargs):
                 variables = self._trigger_variables.async_render(
-                    self.hass, None, limited=True
+                    self.hass,
+                    {"this": this},
```
It's confusing to add the entire state to the trigger variables because the data will be old when the automation triggers.
Maybe just add the `entity_id` here?

emontnemery replied:
On second thought, never mind

emontnemery commented on homeassistant/components/script/__init__.py:
```diff
@@ -401,7 +401,12 @@ async def _async_run(self, variables, context):
             with trace_path("sequence"):
-                return await self.script.async_run(variables, context)
+                states = self.hass.states.get(self.entity_id)
+                this = states.as_dict() if states else {"entity_id": self.entity_id}
```
Same comment as above, when can this be None?

r-t-s replied:
No, but lint complained about states since the return of states.get can be None.

emontnemery commented on tests/components/automation/test_init.py:
```diff
@@ -1273,16 +1273,20 @@ async def test_automation_trigger_variables(hass, caplog):
                     "variables": {
                         "event_type": "{{ trigger.event.event_type }}",
                         "test_var": "overridden_in_config",
+                        "this_var": "{{this.entity_id}}",
```
```suggestion
                        # Verify the this variable is available when rendering variables
                        "this_var": "{{this.entity_id}}",
```

r-t-s replied:
Restructured to test for this during  rendering of variables and rendering of templates in actions.

emontnemery commented on tests/components/automation/test_init.py:
```diff
@@ -1273,16 +1273,20 @@ async def test_automation_trigger_variables(hass, caplog):
                     },
                     "trigger_variables": {
                         "test_var": "defined_in_config",
+                        "this_trig_var": "{{this.entity_id}}",
```
```suggestion
                        # Verify the this variable is available when rendering trigger variables
                        "this_trig_var": "{{this.entity_id}}",
```

r-t-s replied:
Restructured to test for this during rendering of trigger_variables, and rendering of templates in actions.

emontnemery commented on tests/components/script/test_init.py:
```diff
@@ -679,6 +679,7 @@ async def test_script_variables(hass, caplog):
             "script": {
                 "script1": {
                     "variables": {
+                        "this_var": "{{this.entity_id}}",
```
```suggestion
                        # Verify the this variable is available when rendering variables
                        "this_var": "{{this.entity_id}}",
```

r-t-s replied:
Restructured to test for this during rendering of variables, and rendering of templates in actions.

emontnemery commented on homeassistant/components/automation/__init__.py:
```diff
@@ -460,7 +460,11 @@ async def async_trigger(self, run_variables, context=None, skip_condition=False)
             if self._variables:
                 try:
-                    variables = self._variables.async_render(self.hass, run_variables)
+                    states = self.hass.states.get(self.entity_id)
```
Name the variable `state` since it's a single state

r-t-s replied:
Done

emontnemery commented:
@r-t-s Don't to open new PRs if you restart your work locally
Instead, do like this:
1. Rename your old local PR branch (`git -m "r-t-s:this_variable_old"`)
2. Create a new local branch with the name used in the github PR
3. Force-push the new branch, the PR will then automatically be refreshed and point to your new branch

r-t-s commented:
The state can be NULL when running the unit tests.


On Fri, Jul 9, 2021 at 4:19 AM Erik Montnemery ***@***.***>
wrote:

> @r-t-s <https://github.com/r-t-s> Don't to open new PRs if you restart
> your work locally
> Instead, do like this:
>
>    1. Rename your old local PR branch (git -m "r-t-s:this_variable_old")
>    2. Create a new local branch with the name used in the github PR
>    3. Force-push the new branch, the PR will then automatically be
>    refreshed and point to your new branch
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/52774#issuecomment-877045164>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACVV5AI7CL7V5OAY5QU4TZTTW25L5ANCNFSM5ABQ7QGQ>
> .
>


-- 

*Richard T. Schaefer      E-mail: **schaefer@RTS-Services.com*
<schaefer@RTS-Services.com>

*RTS Services Inc            Web: **http://RTS-Services.com*
<http://rts-services.com/>



*1675 Hickory Creek Rd.   ___        Voice:  (512) 923-2755 Marble
Falls, TX 78654    |                        `---------(*)---------'*

r-t-s commented:
The state has the id and friendly_name and can be useful.
But at minimum the entity_id is needed, which is what I originally had
(variable was automation_entity_id)
You recommended I change this to the state, and seem to be recommending
back to entity_id.
Please confirm which you prefer.



On Fri, Jul 9, 2021 at 7:19 AM Richard T Schaefer ***@***.***>
wrote:

> The state can be NULL when running the unit tests.
>
>
> On Fri, Jul 9, 2021 at 4:19 AM Erik Montnemery ***@***.***>
> wrote:
>
>> @r-t-s <https://github.com/r-t-s> Don't to open new PRs if you restart
>> your work locally
>> Instead, do like this:
>>
>>    1. Rename your old local PR branch (git -m "r-t-s:this_variable_old")
>>    2. Create a new local branch with the name used in the github PR
>>    3. Force-push the new branch, the PR will then automatically be
>>    refreshed and point to your new branch
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/home-assistant/core/pull/52774#issuecomment-877045164>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACVV5AI7CL7V5OAY5QU4TZTTW25L5ANCNFSM5ABQ7QGQ>
>> .
>>
>
>
> --
>
> *Richard T. Schaefer      E-mail: **schaefer@RTS-Services.com*
> <schaefer@RTS-Services.com>
>
> *RTS Services Inc            Web: **http://RTS-Services.com*
> <http://rts-services.com/>
>
>
>
> *1675 Hickory Creek Rd.   ___        Voice:  (512) 923-2755 Marble
> Falls, TX 78654    |                        `---------(*)---------'*
>


-- 

*Richard T. Schaefer      E-mail: **schaefer@RTS-Services.com*
<schaefer@RTS-Services.com>

*RTS Services Inc            Web: **http://RTS-Services.com*
<http://rts-services.com/>



*1675 Hickory Creek Rd.   ___        Voice:  (512) 923-2755 Marble
Falls, TX 78654    |                        `---------(*)---------'*

emontnemery commented:
> You recommended I change this to the state, and seem to be recommending
back to entity_id.

I meant specifically for the trigger variables, but never mind.

emontnemery commented:
> The state can be NULL when running the unit tests.

Yes, but do like this instead in `_async_attach_triggers`:
```diff
@@ -589,8 +588,9 @@ class AutomationEntity(ToggleEntity, RestoreEntity):
         variables = None
         if self._trigger_variables:
             try:
-                states = self.hass.states.get(self.entity_id)
-                this = states.as_dict() if states else {"entity_id": self.entity_id}
+                # Make sure the automation's state is in the state machine
+                self.async_write_ha_state()
+                this = self.hass.states.get(self.entity_id).as_dict()
                 variables = self._trigger_variables.async_render(
                     self.hass,
                     {"this": this},
```

In `async_trigger` and `_async_run`, you can just drop the condtion and do:
```py
this = self.hass.states.get(self.entity_id).as_dict()
```

emontnemery commented on homeassistant/components/automation/__init__.py:
```diff
@@ -585,8 +589,12 @@ def log_cb(level, msg, **kwargs):
         if self._trigger_variables:
             try:
+                states = self.hass.states.get(self.entity_id)
+                this = states.as_dict() if states else {"entity_id": self.entity_id}
                 variables = self._trigger_variables.async_render(
-                    self.hass, None, limited=True
+                    self.hass,
+                    {"this": this},
+                    limited=True,
```
This should not be inside the `if self._trigger_variables:`, we should always add the `this` variables:
```diff
-        variables = None
+        # Make sure the automation's state is in the state machine
+        self.async_write_ha_state()
+        this = self.hass.states.get(self.entity_id).as_dict()
+        variables = {"this": this}
         if self._trigger_variables:
             try:
-                states = self.hass.states.get(self.entity_id)
-                this = states.as_dict() if states else {"entity_id": self.entity_id}
                 variables = self._trigger_variables.async_render(
                     self.hass,
-                    {"this": this},
+                    variables,
                     limited=True,
                 )
```

r-t-s replied:
Changed.

emontnemery commented on homeassistant/components/automation/__init__.py:
```diff
@@ -460,7 +460,11 @@ async def async_trigger(self, run_variables, context=None, skip_condition=False)
         ) as automation_trace:
             if self._variables:
```
We should always set the `this` variable:
```diff
+            this = self.hass.states.get(self.entity_id).as_dict()
+            run_variables = {"this": this, **run_variables}
             if self._variables:
                 try:
-                    states = self.hass.states.get(self.entity_id)
-                    this = states.as_dict() if states else {"entity_id": self.entity_id}
                     variables = self._variables.async_render(
-                        self.hass, {"this": this, **run_variables}
+                        self.hass, run_variables
                     )
```

r-t-s replied:
Changed

emontnemery commented:
Add a new test case to make sure the `this` variable is always available, add something similar for script.

```py
async def test_automation_this_variables(hass, caplog):
    """Test automation trigger variables."""
    calls = async_mock_service(hass, "test", "automation")

    assert await async_setup_component(
        hass,
        automation.DOMAIN,
        {
            automation.DOMAIN: [
                {
                    "trigger": {"platform": "event", "event_type": "event_{{ this.entity_id }}"},
                    "action": {
                        "service": "test.automation",
                        "data": {
                            "this_entity_id": "{{ this.entity_id }}",
                        },
                    },
                },
            ]
        },
    )
    hass.bus.async_fire("event_automation.automation_0")
    await hass.async_block_till_done()
    assert len(calls) == 1
    assert calls[0].data.get("this_entity_id") == "automation.automation_0"

    assert "Error rendering variables" not in caplog.text
```

r-t-s commented:
After restructuring I get:
typing run-test: commands[0] | mypy homeassistant
homeassistant/components/automation/__init__.py:592: error: Item "None" of
"Optional[State]" has no attribute "as_dict"  [union-attr]
Found 1 error in 1 file (checked 4054 source files)

For the lines:
        self.async_write_ha_state()
        variables = {"this": self.hass.states.get(self.entity_id).as_dict()}
This is in automation/__init__.py

In  script/__init__.py the following is accepted!
                self.async_write_ha_state()
                return await self.script.async_run(
                    {"this": self.hass.states.get(self
.entity_id).as_dict()},
                    context,
                )

Google Search indicates the following should work:
state = self.hass.states.get(self.entity_id)
assert isinstance(state, State)
Then I should be able to use the following without flagging an error:
{"this": state.as_dict()}

Should I use the same structure in both places ... even though only one is
flagging an error ?

Or maybe just roll back to setting this to be just an entity_id ?

On Fri, Jul 9, 2021 at 9:12 AM Erik Montnemery ***@***.***>
wrote:

> Add a new test case to make sure the this variable is always available,
> add something similar for script.
>
> async def test_automation_this_variables(hass, caplog):
>     """Test automation trigger variables."""
>     calls = async_mock_service(hass, "test", "automation")
>
>     assert await async_setup_component(
>         hass,
>         automation.DOMAIN,
>         {
>             automation.DOMAIN: [
>                 {
>                     "trigger": {"platform": "event", "event_type": "event_{{ this.entity_id }}"},
>                     "action": {
>                         "service": "test.automation",
>                         "data": {
>                             "this_entity_id": "{{ this.entity_id }}",
>                         },
>                     },
>                 },
>             ]
>         },
>     )
>     hass.bus.async_fire("event_automation.automation_0")
>     await hass.async_block_till_done()
>     assert len(calls) == 1
>     assert calls[0].data.get("this_entity_id") == "automation.automation_0"
>
>     assert "Error rendering variables" not in caplog.text
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/52774#issuecomment-877219325>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACVV5ALPRXMCQTVSG5IBA5DTW37WDANCNFSM5ABQ7QGQ>
> .
>


-- 

*Richard T. Schaefer      E-mail: **schaefer@RTS-Services.com*
<schaefer@RTS-Services.com>

*RTS Services Inc            Web: **http://RTS-Services.com*
<http://rts-services.com/>



*1675 Hickory Creek Rd.   ___        Voice:  (512) 923-2755 Marble
Falls, TX 78654    |                        `---------(*)---------'*

emontnemery commented:
You can just do `assert state is not None`, we don't need to check the type.
Only add it where mypy complained.

r-t-s commented:
Update reference to documentation.

r-t-s commented:
I believe I have my repository issues straghtened out.
Addressed all requested issues. 
Fixed issue that showed up on Workflow test, not shown in local testing (dereferencing a variable that was None)

emontnemery commented on homeassistant/components/automation/__init__.py:
```diff
@@ -458,15 +458,19 @@ async def async_trigger(self, run_variables, context=None, skip_condition=False)
             self._trace_config,
         ) as automation_trace:
+
+            self.async_write_ha_state()
```
Is this really needed here, it should only be needed in `_async_attach_triggers`?

r-t-s replied:
I will remove and retest.  I copied the stragegy from _async_attach_triggers ... but it should already have it's state set before  the trigger.

emontnemery commented on homeassistant/components/script/__init__.py:
```diff
@@ -401,7 +401,11 @@ async def _async_run(self, variables, context):
             script_trace.set_trace(trace_get())
             with trace_path("sequence"):
-                return await self.script.async_run(variables, context)
+                self.async_write_ha_state()
```
Is this really needed here, it should only be needed in `_async_attach_triggers`?

r-t-s replied:
Yep, used the same initialization template in all 3 cases ...  I only needed to resolve a failure when incoming variables was None. I will remove and retest.

r-t-s replied:
Unit test passed ... code pushed.

frenck commented on homeassistant/components/automation/__init__.py:
```diff
@@ -445,15 +445,17 @@ async def async_trigger(self, run_variables, context=None, skip_condition=False)
             self._trace_config,
         ) as automation_trace:
+            state = self.hass.states.get(self.entity_id)
+            assert state is not None
+            variables = {"this": state.as_dict(), **(run_variables or {})}
```
asserting in general is not a good idea.
In this case, it is better to set `this` to `None` than having the probability that the `as_dict()` calls throw up a stack trace in the event the state returned `None`.

```suggestion
            this = None
            if state is not None:
                this = state.as_dict()
            variables = {"this": this, **(run_variables or {})}
```

r-t-s replied:
In actuality state will NEVER be null at this point. The assert was added to resolve a lint related issue, because as_dict is not a method of none.  This solution was recommended by Erik Montnemery to replace what I had which was:
      this = states.as_dict() if states else {"entity_id": self.entity_id}

frenck replied:
Right now, it should indeed never be None, however, as things might involve or bugs occur, it still is capable of returning `None` as it is a get operation. We should handle that, not suppress it.

If it ever goes south, it will blow up because it was suppressed and not handled.

r-t-s replied:
Asserts removed. 
Initialized this as:
    this = {"entity_id": self.entity_id" }

frenck replied:
It should be initialized as None: `this = None`
Reasoning: It allows for checking against it. But setting it as you did now, the output will become different all of a sudden.

frenck commented on homeassistant/components/automation/__init__.py:
```diff
@@ -569,11 +571,16 @@ async def _async_attach_triggers(
 
-        variables = None
+        self.async_write_ha_state()
+        state = self.hass.states.get(self.entity_id)
+        assert state is not None
+        variables = {"this": state.as_dict()}
```
```suggestion
        state = self.hass.states.get(self.entity_id)
        this = None
        if state is not None:
            this = state.as_dict()
        variables = {"this": this}
```

r-t-s replied:
This is where I found state to be none in some unit tests  ... I and used my previous solution, similar to yours. But Eric suggested I add the  self.async_write_ha_state() and again use  the assert to get rid of the lint warning.

frenck commented on homeassistant/components/script/__init__.py:
```diff
@@ -401,7 +401,10 @@ async def _async_run(self, variables, context):
             script_trace.set_trace(trace_get())
             with trace_path("sequence"):
-                return await self.script.async_run(variables, context)
+                state = self.hass.states.get(self.entity_id)
+                assert state is not None
+                script_vars = {"this": state.as_dict(), **(variables or {})}
```
```suggestion
                state = self.hass.states.get(self.entity_id)
                this = None
                if state is not None:
                    this = state.as_dict()
                script_vars = {"this": this, **(variables or {})}
```

r-t-s replied:
Same comment as automation run_trigger

r-t-s replied:
Same comment as run_trigger

frenck commented:
@r-t-s Will you please stop generating notifications, rebases, pushes, requests for reviews and other stuff. The queue is big, some patience is needed at times.

As a matter of fact, I was just reviewing this PR and started a CI, which you have now aborted again with your force push.

r-t-s commented:
Sorry, I saw the rebase request on documentation release ... and was do this for all.

frenck commented on homeassistant/components/automation/__init__.py:
```diff
@@ -445,15 +445,19 @@ async def async_trigger(self, run_variables, context=None, skip_condition=False)
             trigger_context,
             self._trace_config,
         ) as automation_trace:
+            this = {"entity_id": self.entity_id}
```
```suggestion
            this = None
```

r-t-s replied:
Changed

frenck commented on homeassistant/components/automation/__init__.py:
```diff
@@ -569,11 +573,18 @@ async def _async_attach_triggers(
             self._logger.log(level, "%s %s", msg, self.name, **kwargs)
 
-        variables = None
+        this = {"entity_id": self.entity_id}
```
```suggestion
        this = None
```

r-t-s replied:
Changed

frenck commented on homeassistant/components/script/__init__.py:
```diff
@@ -401,7 +401,12 @@ async def _async_run(self, variables, context):
             script_trace.set_trace(trace_get())
             with trace_path("sequence"):
-                return await self.script.async_run(variables, context)
+                this = {"entity_id": self.entity_id}
```
```suggestion
                this = None
```

r-t-s replied:
Change

frenck reviewed (approved):
Thanks, @r-t-s 👍 

Nice addition to our automations and scripts engine 🎉
